### PR TITLE
fix cannot reconnect

### PR DESCRIPTION
**Pull request by alonelaval**
closes https://github.com/apache/skywalking/issues/5732

### CODE REVIEW

**Comment by kezhenxu94** [comment]
As discussed by direct messages, we prefer checking the reason why it throws (agent still sends data when the server side closes the socket), I'll take a detailed look soon

**Comment by alonelaval** [comment]
> As discussed by direct messages, we prefer checking the reason why it throws (agent still sends data when the server side closes the socket), I'll take a detailed look soon

ok

File: skywalking/agent/__init__.py [hunk 1]
```diff
@@ -32,18 +32,26 @@
...
-            __protocol.heartbeat()
-
-        __finished.wait(30 if connected() else 3)
+            try:
+                __protocol.heartbeat()
+                __finished.wait(3)
+            except Exception as e:
+                pass
+        else:
+            connect()
```
Comment (kezhenxu94) [inline skywalking/agent/__init__.py:41]: There is another problem when the client **IS CONNECTING**, we should not `connect` again

**Comment by xdrenjj** [comment]
@alonelaval This merge  request is to solve this problem? 
 File "/usr/local/lib/python3.8/asyncio/events.py", line 81, in _run
[ RequestID: ]     self._context.run(self._callback, *self._args)
[ RequestID: ]   File "/data0/codes/test_sw/venv/lib/python3.8/site-packages/tornado/web.py", line 2329, in <lambda>
[ RequestID: ]     fut.add_done_callback(lambda f: f.result())
[ RequestID: ]   File "/data0/codes/test_sw/venv/lib/python3.8/site-packages/skywalking/plugins/sw_tornado.py", line 66, in _sw_get_response
[ RequestID: ]     peer = request.connection.stream.socket.getpeername()
[ RequestID: ] AttributeError: 'NoneType' object has no attribute 'getpeername'

**Comment by kezhenxu94** [comment]
> @alonelaval This merge request is to solve this problem?
> File "/usr/local/lib/python3.8/asyncio/events.py", line 81, in _run
> [ RequestID: ] self._context.run(self._callback, *self._args)
> [ RequestID: ] File "/data0/codes/test_sw/venv/lib/python3.8/site-packages/tornado/web.py", line 2329, in 
> [ RequestID: ] fut.add_done_callback(lambda f: f.result())
> [ RequestID: ] File "/data0/codes/test_sw/venv/lib/python3.8/site-packages/skywalking/plugins/sw_tornado.py", line 66, in _sw_get_response
> [ RequestID: ] peer = request.connection.stream.socket.getpeername()
> [ RequestID: ] AttributeError: 'NoneType' object has no attribute 'getpeername'

No, can you please open another issue to report your problems? And produce more details to help us to solve the problems, such as the tornado version, a minimal demo project, etc.

**Comment by xdrenjj** [comment]
> > @alonelaval This merge request is to solve this problem?
> > File "/usr/local/lib/python3.8/asyncio/events.py", line 81, in _run
> > [ RequestID: ] self._context.run(self._callback, *self._args)
> > [ RequestID: ] File "/data0/codes/test_sw/venv/lib/python3.8/site-packages/tornado/web.py", line 2329, in
> > [ RequestID: ] fut.add_done_callback(lambda f: f.result())
> > [ RequestID: ] File "/data0/codes/test_sw/venv/lib/python3.8/site-packages/skywalking/plugins/sw_tornado.py", line 66, in _sw_get_response
> > [ RequestID: ] peer = request.connection.stream.socket.getpeername()
> > [ RequestID: ] AttributeError: 'NoneType' object has no attribute 'getpeername'
> 
> @xdrenjj No, can you please open another issue to report your problems? And produce more details to help us to solve the problems, such as the tornado version, a minimal demo project, etc.

@kezhenxu94 thanks for your reply. The problem I described may be caused by high concurrency or network rate. By the way, when is the plan to release a new version that solves the can't reconnect problem?







。

**Comment by kezhenxu94** [comment]
Closing in favor of #79 

> By the way, when is the plan to release a new version that solves the can't reconnect problem?

@xdrenjj we plan to release a version after #79 gets merged, please subscribe to our mailing list to stay tune.

**Pull request closed without merging**